### PR TITLE
Add ReadOptions.auto_prefix_mode

### PR DESCRIPTION
**Pull request by siying**
Summary:
Add a new option ReadOptions.auto_prefix_mode. When set to true, iterator should return the same result as total order seek, but may choose to do prefix seek internally, based on iterator upper bounds. Also fix two previous bugs when handling prefix extrator changes: (1) reverse iterator should not rely on upper bound to determine prefix. Fix it with skipping prefix check. (2) block-based filter is not handled properly.

Test Plan: (1) add a unit test; (2) add the check to stress test and run see whether it can pass at least one run.

### CODE REVIEW

**Comment by siying** [comment]
@ltamasi that's a good point. I will modify DBIter and treat auto_prefix_mode the same as total_order_seek.

File: table/block_based/block_based_table_reader.h [hunk 1]
```diff
@@ -724,7 +724,13 @@ class BlockBasedTableIterator : public InternalIteratorBase<TValue> {
            block_iter_points_to_real_block_;
   }
 
-  bool CheckPrefixMayMatch(const Slice& ikey) {
+  bool CheckPrefixMayMatch(const Slice& ikey, bool is_forward_direction) {
```
Comment (riversand963) [inline table/block_based/block_based_table_reader.h]: Change `bool is_forward_direction` to `Direction direction` and compare it with kForward or kReverse?

**Comment by ltamasi** [comment]
> @ltamasi that's a good point. I will modify DBIter and treat auto_prefix_mode the same as total_order_seek.

Actually, it seems to me that `auto_prefix_mode` is currently meant to be used with `total_order_seek==false`? I'm wondering if it would make sense to change that so it's to be used in conjunction with `total_order_seek==true`, considering that the results are supposed to be the same? (That is, `auto_prefix_mode` would be a "submode" or optimized version of `total_order_seek`.) In that case, `DBIter` and friends probably would not need to be modified.

**Comment by siying** [comment]
@ltamasi i hope `auto_prefix_mode` can return the same result as `total_order_seek`. Right now it is meant to be used with `total_order_seek==false` but after the feature is stable, we may consider to make it work with `total_order_seek==true` too. I'll try to make changes in all the places.

**Comment by siying** [comment]
Although it shows that "Facebook Internal - Builds & Tests" failed, inside the link it says the tests finished. Tests passed except the two diffcoverage ones, which are not related to the PR.

File: table/block_based/block_based_table_reader.cc [hunk 2]
```diff
@@ -3202,6 +3204,7 @@ InternalIterator* BlockBasedTable::NewIterator(
     size_t compaction_readahead_size) {
   BlockCacheLookupContext lookup_context{caller};
   bool need_upper_bound_check =
+      read_options.auto_prefix_mode ||
```
Comment (ohhzj) [inline table/block_based/block_based_table_reader.cc:3207]: When auto_prefix_mode=true and PrefixExtractorChanged=false ,user set rep_->index_type == BlockBasedTableOptions::kHashSearch, then user will not to able to use prefix seek. It seems to me that auto_prefix_mode and kHashSearch can't be used at the same time, Can you help me explain why?

**Pull request closed without merging**